Library/Homebrew/extend/os/linux/software_spec.rb

```diff
@@ -1,6 +1,40 @@
-# typed: strict
+# typed: true
 # frozen_string_literal: true
 
+class SoftwareSpec
+  extend T::Sig
+
+  GNU_GCC_REGEXP = /^(gcc)$|^(gcc@(4\.9|[5-9]|[10-99]))$/.freeze
+
+  sig { params(spec: T.any(String, T::Hash[T.untyped, T.untyped])).void }
+  def remove_gnu_compilers(spec)
+    spec = spec.keys.first if spec.is_a?(Hash)
```

Comment:
What are the possible `keys` values? `keys.first` feels like a pretty big smell to be considering they aren't order-dependent?

Reply:
```
String: gcc

Hash: {"gcc"=>:build}

Hash: {"gcc"=>[:build, :test]}
```

This is what I found out using brew readall. Ideally we could create a spec class, with a "formula_name" method, so that we do not have to guess this based on the type. I think it is safe to use the first key here as there will only be a single key-value pair.

Reply:
Cool, I think it's fine in that case to check for `keys.first` as long as an error is raised if there's multiple keys. Given the above, though, it might not be necessary.